Feature/hw 10/create starter

DrEremin opened the pull request:
Создание стартера логирования запросов и ответов

DrEremin committed:
HW-10: перенос логирования в стартер

Добавлена аннотация LogExecutionTime;
Добавлен аспект LogExecutionAspect;
Добавлен EndpointExclusionFlagQualifier;
Добавлен AppProperties;
Добавлен RequestDataFormatter;
Добавлен WebLoggingFilter;
Добавлен WebLoggingRequestBodyAdvice;
К классу LoggingStarterAutoConfiguration добавлена автоконфигурация;
Добавлены зависимости на стартеры web и aop.b

DrEremin committed:
HW-10: Добавлен класс PropertiesCutterFromBody

Добавлен класс PropertiesCutterFromBody;
В проект добавлена публикация этого стартера.

DrEremin committed:
HW-10: удалены версии 1.0 и 1.3 из проекта

Удалены версии 1.0 и 1.3 logging-starter из проекта;
Класс PropertiesCutterFromBody перенесен в пакет util;
Добавлена версия 1.4 logging-starter в проект.

DrEremin committed:
HW-10: удалены версии 1.0 и 1.3 из проекта

Удалены версии 1.0 и 1.3 logging-starter из проекта;
Класс PropertiesCutterFromBody перенесен в пакет util;
Добавлена версия 1.4 logging-starter в проект.

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/LoggingStarterAutoConfiguration.java:
```diff
@@ -1,9 +1,50 @@
-        System.out.format("Печать из библиотеки logging-starter: %s%n", s);
+    @Bean
+    @ConditionalOnProperty(prefix = "logging", value = "log-exec-time", havingValue = "true")
+    LogExecutionAspect logExecutionAspect() {
```
public

DrEremin replied:
Исправил

panyukovnn replied:
`+`

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/LoggingStarterAutoConfiguration.java:
```diff
@@ -1,9 +1,50 @@
     }
 
+    @Bean
+    AppProperties appProperties() {
```
неудачное название для настроек, стартер может импортироваться во множество других микросервисов и его классы должны отражать принадлженость к этому стратреу, в данном случае пододйте LoggingStarterProperties

DrEremin replied:
Исправил

panyukovnn replied:
`+`

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/LoggingStarterAutoConfiguration.java:
```diff
@@ -1,9 +1,50 @@
+
+    @Bean
+    PathMatcher pathMatcher() {
+        return new AntPathMatcher();
```
не надо создавать матчер в качестве бина, можно просто как статическое поле 
 
 private static final AntPathMatcher

DrEremin replied:
Исправил

panyukovnn replied:
`+`

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/exclusion/EndpointExclusionFlagQualifier.java:
```diff
@@ -0,0 +1,29 @@
+    @Autowired
+    private PathMatcher pathMatcher;
+
+    private boolean exclusion;
```
так работать не будет, у тебя здесь хранится глобальная переменная, при изменении ее из разных потоков все сломается

лучше пока убери этот класс, когда мы пройдем лекции по многопоточности сможешь оптимизировать (после 16ой лекции), до этого времени лучше пусть несколько раз проверяется стоит ли логгировать данный эндпоинт

DrEremin replied:
Этот класс удалил. Но вынес эту логику проверки в утилитный класс URIPathMatcherWithPathPatterns.

panyukovnn replied:
`+`

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/util/PropertiesCutterFromBody.java:
```diff
@@ -0,0 +1,30 @@
+
+public class PropertiesCutterFromBody {
+
+    public static String cutProperties(String body, List<String> bodyPaths) {
```
мы не удаляем из логов записи, а маскируем, т.е. вместо значения подставляем звездочки ***

DrEremin replied:
Исправил

panyukovnn replied:
`+`

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/util/PropertiesCutterFromBody.java:
```diff
@@ -0,0 +1,30 @@
+
+            newBody = documentContext.jsonString();
+        } catch (UnsupportedOperationException | IllegalArgumentException ignore) {
+        }
```
надо залоггировать ошибку

DrEremin replied:
Исправил

panyukovnn replied:
`+`

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/util/PropertiesCutterFromBody.java:
```diff
@@ -0,0 +1,30 @@
+                try {
+                    documentContext = documentContext.delete(path);
+                } catch (PathNotFoundException ignore) {
+                }
```
и тут тоже логгировать

DrEremin replied:
Исправил

panyukovnn replied:
Я ошибся, логгировать тут не стоит

DrEremin replied:
Вернул обратно

panyukovnn replied:
`+`

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/webfilter/WebLoggingRequestBodyAdvice.java:
```diff
@@ -0,0 +1,64 @@
+        String headers = RequestDataFormatter.formatRequestHeaders(request, appProperties.getHeaders());
+
+        try {
+            String jsonBody = objectMapper.writeValueAsString(body);
```
вот эту логику вынести бы в PropertiesCutterFromBody

DrEremin replied:
Логику конвертации Object -> jsonString вынес. Обработку заголовков оставил отдельной поскольку к телу они не имеют отношения

panyukovnn replied:
`+`

DrEremin committed:
HW-10: коррекция логики фильтрации логов

Все методы создания бинов объявлены как public;
Класс AppProperties переименован в LoggingStarterProperties;
Удален класс EndpointExclusionFlagQualifier;
Добавлен класс URIPathMatcherWithPathPatterns;
Класс PropertiesCutterFromBody переименован в JsonBodyPropertiesMasker;
Логика преобразования объекта тела в строку вынесена в класс
JsonBodyPropertiesMasker;
Операции маскирования свойств тела объекта залогированы;
Версия 1.4 бибилиотеки logging-starter заменена на 1.5.

DrEremin committed:
HW-10: повторная фиксация некоторых изменений

Удален класс EndpointExclusionFlagQualifier;
Добавлен класс URIPathMatcherWithPathPatterns;
Класс PropertiesCutterFromBody переименован в JsonBodyPropertiesMasker;
Почему-то эти изменения не запушились в прошлом коммите, хотя были выделены.

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/util/JsonBodyPropertiesMasker.java:
```diff
@@ -0,0 +1,55 @@
+
+            for (String path : bodyPaths) {
+                try {
+                    documentContext = documentContext.set(path, "***");
```
кажется лучше подойдет функция map

DrEremin replied:
Исправил, однако set смотрелся аккуратнее

panyukovnn replied:
`+`

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/util/JsonBodyPropertiesMasker.java:
```diff
@@ -0,0 +1,55 @@
+                try {
+                    documentContext = documentContext.set(path, "***");
+                } catch (PathNotFoundException ignore) {
+                    log.info("Свойство тела по шаблону '{}' не найдено", path);
```
этот лог убираем

DrEremin replied:
Исправил

panyukovnn replied:
`+`

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/util/JsonBodyPropertiesMasker.java:
```diff
@@ -0,0 +1,55 @@
+            newBody = documentContext.jsonString();
+        } catch (UnsupportedOperationException e) {
+            log.info("Строка тела не соответствует формату JSON");
+        } catch (IllegalArgumentException e) {
```
если null, то тело даже не пробуем маскировать

DrEremin replied:
Исправил

panyukovnn replied:
`+`

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/util/JsonBodyPropertiesMasker.java:
```diff
@@ -0,0 +1,55 @@
+public class JsonBodyPropertiesMasker {
+
+    private static final Logger log = LoggerFactory.getLogger(JsonBodyPropertiesMasker.class);
+    private static final ObjectMapper objectMapper = new ObjectMapper()
```
лучше воспользоваться готовым бином ObjectMapper из spring, т.к. просто заавтовайрить существующий objectMapper

DrEremin replied:
Сделал, но класс JsonBodyPropertiesMasker перестал быть утилитным и из него теперь создается бин, поскольку ObjectMapper не внедрить автовайрингом на статическое поле.

panyukovnn replied:
да, пусть будет бином

`+`

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/util/JsonBodyPropertiesMasker.java:
```diff
@@ -0,0 +1,55 @@
+
+    public static String maskProperties(Object body, List<String> bodyPaths) {
+        String bodyString = body.getClass() != String.class ? convertObjectToJsonString(body) : (String) body;
+        String newBody = "{}";
```
кажется возвращать такое значение - неправильно, т.к. если тело null, то должно и возвращаться null, пустой json - {} и null это все же разные вещи

DrEremin replied:
Теперь метод maskProperties() возвращает Optional и если тело отсутствует, то оно вообще не выводится.

panyukovnn replied:
`+`

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/util/URIPathMatcherWithPathPatterns.java:
```diff
@@ -0,0 +1,26 @@
+
+public class URIPathMatcherWithPathPatterns {
+
+    @Autowired
```
надо убрать Autowired

DrEremin replied:
Исправил

panyukovnn replied:
`+`

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/util/URIPathMatcherWithPathPatterns.java:
```diff
@@ -0,0 +1,26 @@
+
+        for (String pathPattern : pathPatterns) {
+            if (pathMatcher.match(pathPattern, uriPath)) {
+                isMatch = true;
```
можно здесь сразу сделать return true и тогда не понадобиться создавать переменную isMatch и делать break в цикле

DrEremin replied:
Исправил

panyukovnn replied:
`+`

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/webfilter/WebLoggingFilter.java:
```diff
@@ -0,0 +1,53 @@
+            super.doFilter(request, responseWrapper, chain);
+
+            if (!URIPathMatcherWithPathPatterns.matchAny(request.getRequestURI(), loggingStarterProperties.getUriPaths())) {
+                if (method.equals("GET") || method.equals("DELETE")) {
```
даже GET и DELETE методы могут иметь тело, так что эта проверка некорректна

DrEremin replied:
Теперь вместо этого проверка на длину content-length

panyukovnn replied:
это лишнее, не стоит делать проверок никаких, твой метод сам определит надо ли что-то маскировать по переданному телу

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/webfilter/WebLoggingFilter.java:
```diff
@@ -0,0 +1,53 @@
+                }
+
+                String responseBody = new String(responseWrapper.getContentAsByteArray(), StandardCharsets.UTF_8);
+                log.info("Обработка тела ответа...");
```
избыточный лог

DrEremin replied:
Убрал

panyukovnn replied:
`+`

DrEremin committed:
HW-10: исправления обработки тел запросов/ответов

Класс JsonBodyPropertiesMasker перестал быть утилитным и теперь
создается его бин;
Кроме того, он перенесен в пакет masking;
Скорректирован код метода URIPathMatcherWithPathPatterns.anyMatch();
Изменено возвращаемое значение метода
JsonBodyPropertiesMasker.maskProperties();
Исправлено условие определения запроса без тела в методе
WebLoggingFilter.doFilter();
Обновление публикации logging-starter до версии 1.6.

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/masking/JsonBodyPropertiesMasker.java:
```diff
@@ -0,0 +1,66 @@
+            return newBodyOptional;
+        }
+
+        String bodyString = body.getClass() != String.class ? convertObjectToJsonString(body) : (String) body;
```
так класс не проверяют, используется ключевое слово instanceof

panyukovnn commented on src/main/java/ru/dreremin/loggingstarter/masking/JsonBodyPropertiesMasker.java:
```diff
@@ -0,0 +1,66 @@
+            log.info("Не удалось преобразовать объект тела в строку JSON");
+        }
+
+        return jsonString;
```
вот тут кажется некорректно возвращать пустую строку если выбросилась ошибка при преобразовании, лучше вернуть null

DrEremin committed:
HW-10: удаление проверки наличия тела запроса

Удалена проверка наличия тела запроса в методе
WebLoggingFilter.doFilter();
Изменено возвращаемое значение в методе
JsonBodyPropertiesMasker.convertObjectToJsonString().

DrEremin committed:
Revert "HW-10: удаление проверки наличия тела запроса"

This reverts commit 95c8f99232f59844511e38b52dafef3c6359405a.

DrEremin committed:
HW-10: удаление проверки наличия тела запроса

Удалена проверка наличия тела запроса в методе
WebLoggingFilter.doFilter();
Изменено возвращаемое значение в методе
JsonBodyPropertiesMasker.convertObjectToJsonString().

DrEremin committed:
HW-10: возврат проверки наличия тела запроса

DrEremin committed:
HW-10: удалена сборка из проекта

DrEremin committed:
HW-10: добавлена сборка в проект

DrEremin committed:
HW-10: добавлена сборка 1.7 в проект
